### PR TITLE
feat: custom server entry points during development

### PR DESCRIPTION
**Pull request by teemingc**
split off from https://github.com/sveltejs/kit/pull/15574

EDIT: This PR adds a `customHandler` property which adapters can resolve to a module that runs `server.init` and `server.respond` with additional logic. The handler runs during dev, build, prerender, and preview. It's main purpose is to set us up to unify the `dev` and `preview` experiences.

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

File: packages/kit/src/exports/vite/index.js [hunk 1]
```diff
@@ -587,6 +604,111 @@ function kit({ svelte_config }) {
...
+		},
+		load: {
+			filter: {
+				id: [
+					exactRegex('sveltekit:env'),
+					exactRegex(sveltekit_dev_manifest_data),
+					exactRegex(sveltekit_dev_traced)
+				]
+			},
+			handler(id) {
```
Comment (teemingc) [inline packages/kit/src/exports/vite/index.js]: I want to get rid of these virtual modules. It was copied over and adapted from the fetchable dev environment stuff but doesn't need to be a virtual module in the context of this PR.

It should be cleaner to write/return a file in the `resolveId` hook or just pass the `manifest_data` directly to the `ssr_manifest.js` file since we're still running in the same process.

File: packages/kit/src/runtime/app/server/index.js [hunk 2]
```diff
@@ -63,7 +64,7 @@ export function read(asset) {
 		const length = manifest._.server_assets[file];
 		const type = manifest.mimeTypes[file.slice(file.lastIndexOf('.'))];
 
-		return new Response(read_implementation(file), {
+		return new Response(read_implementation(__SVELTEKIT_DEV__ ? from_fs(file) : file), {
```
Comment (teemingc) [inline packages/kit/src/runtime/app/server/index.js:67]: Moved`from_fs` out of the dev read implementation to here so that adapters don't have to repeat `from_fs` when providing a read implementation

File: packages/kit/src/exports/vite/index.js (hunk 1 above)
Comment (teemingc) [reply packages/kit/src/exports/vite/index.js]: done

File: packages/kit/src/exports/vite/dev/index.js [hunk 3]
```diff
@@ -411,11 +106,6 @@ export async function dev(vite, vite_config, svelte_config, get_remotes, root, s
...
 	watch('unlink', () => debounce(update_manifest));
 	watch('change', (file) => {
-		// `manifest_data` is populated lazily on the first request (see `update_manifest`
-		// call in the middleware below), so it may still be undefined if a file changes
-		// before the dev server has served a request. In that case there's nothing to
-		// update — the manifest will be created from scratch on the first request.
-		if (!manifest_data) return;
 		// Don't run for a single file if the whole manifest is about to get updated
 		// Unless it's a file where the trailing slash page option might have changed
 		if (timeout || !/\+(page|layout|server).*$/.test(file)) return;
```
Comment (Nic-Polumeyv) [inline packages/kit/src/exports/vite/dev/index.js:422]: This needs the `if (!manifest_data) return;` guard that used to be here. `update_manifest` catches `sync.create` errors, so with an invalid route tree `manifest_data` stays undefined and the next edit to a `+page`/`+layout`/`+server` file kills the whole process:

```
TypeError: Cannot read properties of undefined (reading 'nodes')
    at Module.update (packages/kit/src/core/sync/sync.js:54:35)
```

Repro: create `src/routes/a/+page.svelte` and `src/routes/(g)/a/+page.svelte`, start dev, request any page (500 as expected), then touch `src/routes/+page.svelte`. With the guard restored the server keeps serving the error page and recovers once the conflict is removed.

File: packages/kit/src/exports/vite/dev/context.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * @typedef {object} Context
+ * @property {ValidatedConfig} svelte_config
+ * @property {ViteDevServer} vite
+ * @property {ManifestData} manifest_data
+ * @property {Array<{ hash: string, file: string }>} remotes
+ */
+
+class DevContext {
+	/** @type {Context} */
+	context = import.meta.hot?.data ?? {};
```
Comment (Nic-Polumeyv) [inline packages/kit/src/exports/vite/dev/context.js]: `import.meta.hot` is undefined when HMR is off (Vite's `createHMROptions` returns false for `server.hmr: false`), so this falls back to a fresh object on every evaluation and the context written in `configureServer` never reaches the server entry. With `server: { hmr: false }` in `vite.config.js` every request 500s with `Cannot read properties of undefined (reading 'kit')`. A module-level fallback object that survives re-evaluation fixes it, though passing the context in explicitly would make the failure impossible rather than handled.

File: packages/kit/src/exports/vite/dev/ssr_manifest.js [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+	routes: dev.manifest_data.routes,
+	params_path: dev.manifest_data.params,
+	root: __SVELTEKIT_ROOT__,
+	load: (file) => loud_ssr_load_module(to_fs(file))
+});
+
+/** @type {SSRManifest} */
+export const manifest = {
+	appDir: __SVELTEKIT_APP_DIR__,
+	appPath: `${__SVELTEKIT_PATHS_BASE__}/${__SVELTEKIT_APP_DIR__}`,
```
Comment (Nic-Polumeyv) [inline packages/kit/src/exports/vite/dev/ssr_manifest.js]: This produces `/_app` where the build produces `_app` (`index.js:1710` strips the leading slash). adapter-node's handler and adapter-cloudflare's worker both prepend `/` to `manifest.appPath`, so in dev they'd look for `//_app/...` and never match. A single shared define used by both here and `build_data.app_path` would keep them from drifting.

File: packages/kit/src/exports/vite/dev/server.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		return await /** @type {InternalServer['respond']} */ (super.respond)(request, {
+			...options,
+			before_handle: async (event, config, prerender, handle) => {
+				// we need to use .run because .enterWith() is not supported in Cloudflare Workers
+				// see https://blog.cloudflare.com/workers-node-js-asynclocalstorage/
+				return await async_local_storage.run({ event, config, prerender }, () => {
+					return options.before_handle?.(event, config, prerender, handle) ?? handle();
+				});
+			},
+			emulator
```
Comment (Nic-Polumeyv) [inline packages/kit/src/exports/vite/dev/server.js]: `emulator` after the spread always overwrites `options.emulator`, so a custom entry can't supply its own platform. `options.emulator ?? emulator` would match how `before_handle` composes above.

File: packages/kit/src/exports/vite/index.js [hunk 7]
```diff
@@ -352,6 +361,7 @@ function kit({ svelte_config }) {
 					: 'globalThis.__sveltekit_dev';
 
 				env = loadEnv(config_env.mode, kit.env.dir, '');
+				server_instrumentation = resolve_entry(path.join(kit.files.src, 'instrumentation.server'));
```
Comment (Nic-Polumeyv) [inline packages/kit/src/exports/vite/index.js]: Creating `src/instrumentation.server.js` mid-session does nothing until a manual restart now that resolution happens here (on version-3 the per-request lookup picked it up on the next request). The watcher in `dev/index.js` that handles the other special files could watch it and call `vite.restart()` when it appears or disappears; re-resolving alone isn't enough because the id is baked into a cached virtual module.

The build also errors when `adapter.supports.instrumentation` is false (line 1415) but this path doesn't check, so an unsupported setup works in dev and first fails at build. Checking here gives both paths the same gate. And the build input hook re-derives this same expression at line 1412 into a shadowing local that could use this one instead.

File: packages/kit/src/runtime/server/index.js [hunk 8]
```diff
@@ -113,16 +113,34 @@ export class Server {
 								return;
 							}
 
+							/** @type {string} */
```
Comment (Nic-Polumeyv) [inline packages/kit/src/runtime/server/index.js]: Living inside the `module.handleError ||` fallback means any app that defines `handleError` gets no overlay, which is most of the apps that would want one. It's also the only Vite-aware code under `runtime/server`. Splitting this into its own PR, run regardless of a user `handleError`, might be easier to review than carrying it here. Two smaller things in the block itself: `message ||= e.message` sits inside `if (e.stack)`, so an error with a message but no stack produces no overlay, and the `vite:` prefix is reserved by Vite for its own events, `sveltekit:error` would avoid collisions.

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 9]
```diff
@@ -66,3 +66,70 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+export async function fetch(request) {
+	return await server.respond(request, {
+		getClientAddress: () => {
+			return request.headers.get('how-your-platform-exposes-the-remote-address')
+		}
+	});
+}
+
+// Without this, server file changes will invalidate the entire Vite server module graph:
+import.meta.hot?.accept();
```
Comment (Nic-Polumeyv) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md]: kit's own `ssr_entry.js` doesn't carry this line (the #15574 version did), and in my testing the log was identical with and without it, `(ssr) page reload` then `(ssr) program reload` either way. Either restore it in `ssr_entry.js` or drop the causal claim here, whichever reflects the intent.

File: packages/kit/src/exports/vite/dev/index.js [hunk 10]
```diff
@@ -602,27 +250,22 @@ export async function dev(vite, vite_config, svelte_config, get_remotes, root, s
...
+					request: req
 				});
 
+				const runner = get_runner(vite);
+
+				/** @type {{ fetch(request: Request): Promise<Response> }} */
+				const server_entry = await runner.import(sveltekit_dev_init);
+
+				if (req.socket.remoteAddress) {
+					request.headers.set('x-sveltekit-remote-address', req.socket.remoteAddress);
```
Comment (Nic-Polumeyv) [inline packages/kit/src/exports/vite/dev/index.js]: When `req.socket.remoteAddress` is falsy a client-supplied `x-sveltekit-remote-address` passes straight through to `getClientAddress()`. Deleting the header unconditionally before the conditional set closes that. It's also now part of the contract every custom entry has to implement, so it deserves a mention in the adapter docs where the example currently says `how-your-platform-exposes-the-remote-address`.

File: packages/kit/src/types/ambient.d.ts [hunk 11]
```diff
@@ -146,3 +146,56 @@ declare module '$app/types' {
...
+ * import { env } from 'sveltekit:env';
+ * import { Server } from 'sveltekit:server';
+ * import { manifest } from 'sveltekit:server-manifest';
+ *
+ * const server = new Server(manifest);
+ *
+ * await server.init({ env });
+ * ```
+ */
+declare module 'sveltekit:server' {
```
Comment (Nic-Polumeyv) [inline packages/kit/src/types/ambient.d.ts]: These ids only resolve under `apply: 'serve'` with `consumer === 'server'`, so the types promise modules that don't exist in a build, and a client-side import gets Vite's raw resolve error instead of kit's server-only message. Meanwhile `sveltekit:server-entry`, the one id users actually reference, isn't declared.

**Review by Nic-Polumeyv — COMMENTED** [review]
I ran this branch against `test/apps/dev-only` and went looking for edge cases. Two of them take the dev server down entirely and one breaks the adapter handlers this is meant to make reusable, details inline with repros.

While testing I ended up writing the test this needs anyway, a plugin in the test app's `vite.config.js` that resolves `sveltekit:server-entry` to a local module setting a marker header, plus one assertion in the dev-only suite. I can PR it against your branch if you want it.

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 12]
```diff
@@ -66,3 +66,70 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+// @errors: 1005 1128
+{
+	name: 'vite-plugin-name-it-yourself',
+	applyToEnvironment(environment) {
+		return environment.name === 'ssr';
+	},
+	resolveId: {
+		filter: {
+			id: /^sveltekit:server-entry$/
+		},
```
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md]: This API is very indirect — if this is something we intend to allow adapter authors to do (and it is) then it should be a first-class API — something like this, perhaps:

```diff
export default function adapter() {
  return {
    name: 'my-adapter',
    async adapt(builder) {
      // ...
    },
+   customHandler: import.meta.resolve('./handler.js')
  };
}
```

Naming is bikesheddable but you get the idea. I didn't use the word 'dev' in the name because I don't see any reason we wouldn't want to use this in `vite preview` as well.

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 13]
```diff
@@ -66,3 +66,70 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+	}
+}
+```
+
+This module should instantiate the server with your app's manifest, initialise environment variables and the `read` implementation, and export a `fetch` handler which receives a `Request` and returns a `Response`:
+
+```js
+import { env } from 'sveltekit:env';
+import { Server } from 'sveltekit:server';
+import { manifest } from 'sveltekit:server-manifest';
```
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md]: The top of the thread already alludes to this but I definitely think we can get rid of these magic imports. If the custom handler is _given_ a server instance instead of having to create one, then presumably we don't need to think about `env` or `read` for example. We can even get rid of the need to import `server` if we make the signature more like this:

```js
export default (server) => {
  return (request) => {
    if (new URL(request.url).pathname === '/_special') {
      return new Response('intercepted');
    }

    return server.respond(request, {
      getClientAddress: () => {...}
    });
  };
};
```

(Not saying it should be exactly this but you get the idea.)

> My gut feeling is that it was originally designed that way to make generating split serverless functions easier as different `manifest`s can be injected into each function entrypoint while reusing the same `Server` class import from the build output

I think we can iterate on that stuff too but to my mind it's independent of the custom entrypoint stuff. You likely wouldn't want to reuse logic between this handler and your production code — for example `/_special` in the example above is something that the platform would handle before your code even ran (think `/_vercel/image`, etc)

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md (hunk 9 above)
Comment (Rich-Harris) [reply documentation/docs/25-build-and-deploy/99-writing-adapters.md]: It feels like an implementation detail that adapter authors shouldn't need to worry about

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md (hunk 12 above)
Comment (Rich-Harris) [reply documentation/docs/25-build-and-deploy/99-writing-adapters.md]: (to be clear this would be part of the Vite app so it would be able to use `$app/...` imports etc)

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md (hunk 13 above)
Comment (Rich-Harris) [reply documentation/docs/25-build-and-deploy/99-writing-adapters.md]: To expand on 'iterate on that stuff', here's where my head is currently at: I think we can avoid exposing `manifest` _at all_ to adapter code — rather than having `builder.generateManifest(...)` we could have something like `builder.generateServer(...)` that creates a module like this:

```js
export const server = new Server({
  // ...the manifest...
});
```

The adapter's job is then to create one of these modules for each group of routes, write them to disk wherever it pleases, and use it in the platform glue code:

```js
const server = `${tmp}/${group.name}/index.js`;
const dest = `.my-platform/functions/${group.name}/index.js`;

builder.generateServer(server, {
  routes: group.routes
});

fs.writeFileSync(dest, `
import { server } from '${path.relative(path.dirname(dest), server)}';

server.init({
  env: MyPlatform.getEnv()
});

export default {
  fetch(request, ctx) {
    return server.respond(request, {
      getClientAddress: () => ctx.clientAddress
    });
  }
};
`);
```

Doing this frees us up to treat the manifest as evolvable implementation detail rather than public API that needs to be maintained, and is just all around a bit neater. (Some adapters _do_ use `manifest` in their code but not in a way that changes anything. Making the contents of `$app/manifest` available to adapters would suffice.)

**Review by Rich-Harris — CHANGES_REQUESTED** [review]
Was holding off on submitting a review for this until I felt like I had a firm grasp of what shape things should take. But that's taking longer than I hoped, so... here you go!

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md (hunk 13 above)
Comment (Nic-Polumeyv) [reply documentation/docs/25-build-and-deploy/99-writing-adapters.md]: In dev the handler only sees a `Request` and the socket is in kit's middleware, so `getClientAddress` needs the address handed in from outside. Passing it to the handler along with the request would avoid a header contract like `x-sveltekit-remote-address`.

Just fyi for teemingc

Comment (Rich-Harris) [reply documentation/docs/25-build-and-deploy/99-writing-adapters.md]: that's true, a better example would have been `platform`

Comment (teemingc) [reply documentation/docs/25-build-and-deploy/99-writing-adapters.md]: Slightly off-topic, I was hoping to deprecate `platform` at some point since platforms such as Vercel, Cloudflare, and Netlify are moving towards making these importable/global rather than a reference passed to the request handler

File: packages/kit/src/exports/vite/dev/server.js (hunk 6 above)
Comment (teemingc) [reply packages/kit/src/exports/vite/dev/server.js]: Mm I think it's better if we introduce a public API to help with the prerendering in dev stuff rather than me adding this hacky workaround. I'm going to change the type signature so that we only accept publicly documented `options` here

File: packages/kit/src/runtime/server/index.js (hunk 8 above)
Comment (teemingc) [reply packages/kit/src/runtime/server/index.js]: `import.meta.hot` gets treeshaken during the build so it's fine to do Vite stuff in the runtime code (it's the same in `client.js`). It also needs to be part of this PR because it's only useful once we start running the server in the environment rather than outside it

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 14]
```diff
@@ -66,3 +68,60 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+};
+```
+
+The handler file should export a default function that returns a request handler. The should return a `Response` such as by calling `server.respond` along with any [platform-specific context](types#Platform). Optionally, you can pass a different set of environment variables or `read` implementation to `server.init` if your platform supports it.
+
+```js
+/// file: src/handler.js
+// @errors: 2322
+/** @type {import('@sveltejs/kit').SSRHandler} */
+export default async function handler(server, env) {
```
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md]: When would adapters need to worry about the `env`? We can just set it directly from Vite surely — even in a fetchable dev environment world, we can still do that init work before passing the server to the adapter's custom handler

```suggestion
export default async function handler(server) {
```

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 15]
```diff
@@ -66,3 +68,60 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+
+	// this is optional. Since `server.init` only runs once, the default
+	// implementation will be used instead if it's not been called already
+	await server.init({
+		env,
+		read: (file) => {
+			// how your platform reads files
+		}
+	});
+
```
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md]:
```suggestion
```

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 16]
```diff
@@ -66,3 +68,60 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+	return async (request, options) => {
```
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md]: This feels like an unnecessary abstraction leak

```suggestion
	return async (request) => {
```

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 17]
```diff
@@ -66,3 +68,60 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+		// custom request/response handling logic goes here
+
+		return await server.respond(request, {
+			...options,
```
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md]:
```suggestion
```

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 18]
```diff
@@ -66,3 +68,60 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+	// implementation will be used instead if it's not been called already
+	await server.init({
+		env,
+		read: (file) => {
+			// how your platform reads files
+		}
+	});
+
+	return async (request, options) => {
+		// custom request/response handling logic goes here
```
Comment (Rich-Harris) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md:108]: would be nice to include an example, however simple (in fact simpler probably === better)

**Review by Rich-Harris — CHANGES_REQUESTED** [review]
I think we can simplify this a fair chunk, both API-wise (as mentioned above, making the adapter author deal with `env` and `read` etc feels unnecessary) and implementation-wise (rather than having a new plugin, we can just consistently have an entry for the handler that's basically `customHandler ?? `${...}/handler.js`. Opened #16560 as a starting point

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md (hunk 14 above)
Comment (teemingc) [reply documentation/docs/25-build-and-deploy/99-writing-adapters.md]: Yeah, it was only included because right now you have to include `env` and `read` together when calling `server.init`

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md [hunk 19]
```diff
@@ -66,3 +68,50 @@ Within the `adapt` method, there are a number of things that an adapter should d
...
+
+```js
+/// file: src/handler.js
+// @errors: 2322
+/** @type {import('@sveltejs/kit').SSRHandler} */
+export default async function handler(server) {
+	// perform setup work here
+
+	return async (request) => {
+		// custom request/response handling logic goes here
```
Comment (teemingc) [inline documentation/docs/25-build-and-deploy/99-writing-adapters.md:108]:
```suggestion
		// custom request/response handling logic goes here

		if (new URL(request.url).pathname === '/api') {
			return fetch('https://example.com/api');
		}
```
Maybe something like this to proxy requests without invoking SvelteKit logic?

File: packages/kit/src/exports/vite/preview/index.js [hunk 20]
```diff
@@ -209,24 +213,30 @@ export async function preview(vite, vite_config, svelte_config) {
...
+							const { remoteAddress } = req.socket;
+							if (remoteAddress) return remoteAddress;
+							throw new Error('Could not determine clientAddress');
+						},
+						read: (file) => {
+							if (file in manifest._.server_assets) {
+								return fs.readFileSync(join(dir, file));
+							}
+
+							return fs.readFileSync(join(svelte_config.kit.files.assets, file));
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/preview/index.js]: One way this could presumably fail is if the custom handler points at user code (i.e. the adapter gives users the ability to specify a custom handler, rather than merely providing one), in which case presumably changes to that code would not be reflected in the module graph.

I think we might instead need to have an alias that points at the handler module, and instead of importing the handler directly we import a module that imports that alias, does the initialization, and exports a `respond` function. That way, the initialization will only happen once under normal circumstances, but will happen again whenever the relevant part of the module graph is invalidated.

File: packages/kit/src/core/postbuild/prerender.js [hunk 21]
```diff
@@ -301,30 +305,37 @@ async function prerender({ hash, out, manifest_path, metadata, verbose, env, vit
...
-				// Static assets emitted during build
-				if (file.startsWith(config.appDir)) {
-					return readFileSync(`${out}/server/${file}`);
-				}
-
-				// stuff in `static`
-				return readFileSync(join(config.files.assets, file));
-			},
-			emulator
+		const respond = await init_server({
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/core/postbuild/prerender.js]: See https://github.com/sveltejs/kit/pull/16464#discussion_r3677616455 — the same applies here, I think

File: documentation/docs/25-build-and-deploy/99-writing-adapters.md (hunk 19 above)
Comment (Rich-Harris) [reply documentation/docs/25-build-and-deploy/99-writing-adapters.md:108]: This is app-specific code, it wouldn't appear in an adapter's generic custom handler. I think we can be super abstract here:

```suggestion
		// custom request/response handling logic goes here
		if (new URL(request.url).pathname === '/_ping') {
			return new Response('pong');
		}
```

File: packages/kit/src/runtime/app/server/index.js (hunk 2 above)
Comment (Rich-Harris) [reply packages/kit/src/runtime/app/server/index.js:67]: Slightly tangential to this PR but something I noticed earlier is that `from_fs` isn't quite sufficient — if someone is (for example) using `enhanced-img` then their import images will _not_ correspond to files on disk during dev:

```js
import potato from './potato.jpg?url';

console.log(potato); // '/@imagetools/c85bbc2b12afef1f492149e7f5fb432b4c3feac8'

// throws 'Error: Asset does not exist: @imagetools/c85bbc2b12afef1f492149e7f5fb432b4c3feac8'
const asset = read(potato); 
```

Probably better to tackle in a follow-up though

File: packages/kit/src/exports/vite/preview/index.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/preview/index.js]: Ah right, I missed that this was preview and not dev code. Fair enough. I think we could punt on the alias stuff for now, in the name of shipping

**Review by Rich-Harris — APPROVED** [review]
fantastic work!

File: packages/kit/src/runtime/app/server/index.js (hunk 2 above)
Comment (teemingc) [reply packages/kit/src/runtime/app/server/index.js:67]: Filed an issue for it https://github.com/sveltejs/kit/issues/16585
